### ISSUE TITLE
Suggestion: Multi quality of life features.

### ISSUE THREAD

**Issue by Raizanad** [RimSort/RimSort#406]
Problem: User SteamDB creation is unrealiable.
Solution: Update the community SteamDB library more frequently via automation.

Problem: The game reads and logs inactive mods. It is burdensome for bug fixing and mod maintainance/fine tuning.
Solution: Vortex method - Symlink creation for active mods to "RimWorld/Mods"
Example: Mod 32932124 is set as active then generate symlink into the Mods folder.

Problem: There is no point asking the user to set the (Game) Mods folder location if it is always the same.
Solution: Automatically acquire it when root folder is set.
Additional: Leave the Mods folder field for where the user wants to have their mod collection/staging folder. Should default to "RimSort/User/Mods".
The Steam file mod location has a purpose, but it can also be used for the database/local/base mod folder of already installed mods that were not directly downloaded by RimSort SteamCMD.

Problem: Deleting/moving mods is a nightmare with Steam coded numbering but no steam subscription user interface.
Solution: Upon SteamCMD mod download make it so the name of the staging folder is "21232141 The Mod Name" or "The Mod Name" while the symlink generated into "RimWorld/Mods" upon activating it can have the Steam Workshop coding.

Problem: Automatically update mods on refresh is a pain.
Solution Update mods automatically every 12h, 24h, daily, every 3 days, weekly and so on.

Problem: Search for RimSort mod update automatically on startup shouldn't be on "Help"
Solution: Move it to settings. Also add the option for once a day instead on every start.

Problem: When the app crashes, the download list goes with it.
Solution: Make a backup of the last operation and check if it was completed successfully, if not then on app startup ask the user if he wants to continue.

Problem: Control + c doesn't automatically work on the app.
Solution: Enable it instead of right clicking to do it.

Problem: SteamDB out of date spam.
Solution: No warning if the steam DB is outdated by a certain period (whatever usually takes to update the github community one).

Problem: The app waits for the SteamDB popup or the duplicate mod popup to continue operations.
Solution: The app shouldn't wait for this sort of warning to be acknowledged since they are minor. It should continue the rest of the startup.

Problem: Closing the app too fast after opening if the popup is going to show makes the app not close properly and need task manager to force close it.
Solution: A fix.

Problem: Inputting a wrong collection link or cancelling makes a double popup to trigger instead of one.
Solution: If the link is invalid the popup (once) is fine, but if the field was empty no need to warn the user.

Problem: Updating community rules and community database hidden away.
Solution: Automatization once a day checking the github.

Problem: Searching keywords needs to be too precise.
Solution: When searching mods in the searchbox making skipping words fine. I.e: Vanilla Expanded will still yield results no matter if you skip the middle keyword.

Feature: By size of mod sorting for the mods in the app (not the mod list final).
Feature: Alphabetical sorting for the mods in the app (not the mod list final).
Feature: Show to the user when the mod was installed/downloaded.
Feature: Show to the user when the mod was last updated.
Feature: Show to the user when the mod was moved to the active list and sort by this.
Feature: Sort to the top mods that have errors and warnings.

Problem: Too many mods fail to upload to Rentry.co.
Solution: A similar export function to rentry.co but as notepad txt file for the user to handle it. The export to clipboard is too different than the Rentry one. Both have value.

Problem: Options for multiple selected mods are limited compared to single selection.
Solution: Selecting multiple mods then right clicking should open multiple links to steam instead of just hiding the option (for example).

Problem: Toggle warning off should eliminate the warning and change the color of the text instead of just removing the little icon.
Solution: Remove it from the warnings at the bottom of the app, and change the color from red/orange from the text to a purple to signify user manually set the mod as fine. Many mods for 1.4 work on 1.5, specially textures as an example. Would be great if it changed the about.xml too for the game not to trigger errors but this may be too much since you will need a db to remove the change (or changing about.xml to about.xml.old).

Problem: Interface is a bit weird.
Solution: Move the image and description of the mod to the center, and set a limit to the sizing for the image since some mods have really big images that widen the column.
More: 3 columns: Left = inactive mods, center = description, right = active mods.

Feature: In settings add the option for just one search bar to search both fields at the same time. The current method is great, but sometimes you are just looking for things other modders tell you are broken for example.

Problem: Trying to validate all mods opens steam and does nothing.
Solution: Use SteamCMD.

Problem: App currently deletes all mods that you want to redownload, then queues them in SteamCMD. This is a problem if the app crashes or if the mod is no longer available.
Solution: When redownloading mods, leave the main folder in case it fails or the app crashes so it can be queued again. Would be great to check if the mod is still available and in case it was taken down then skip it and warn the user.

Problem: Unclear behavior of: Instances, Sort mods Topologically, Dry run todds are for... Even when looking at todds github for example.
Solution: Tooltip or something.

Problem:: SteamCMD must be purged and redownloaded every other day because it gets faulty somehow.
Solution: ""\steam\steamapps\workshop\appworkshop_294100.acf"" should be removed periodically since it eventually causes errors, disconnections and steam refusals for mods downloading.

Problem: Currently RimSort installs itself into too many places. AppData/Local, AppData/Local/Temp, AppData/LocalLow, wherever the user installed the portable app, and uses RimWorld/Mods.
Solution: Since there is no "auto clean rimsort" button. Use the Portable app root folder to create a User folder that isn't removed on updates of the app. User can contain User/Steam for SteamCMD and User/Instances, User/Databases, User/Collection or User/Lists, User/Settings and User/Mods.

Problem: Technically this is good, but keep your account logged in the built in browser since some mods can't be seem without logging due to censorship. Also may helps with downloading from subscriptions.
Solution: Keep cookies for some time.

Feature: Subscription list based mod downloading, and user/workshop downloading for when you want everything made by a certain modder.
Feature: Add a button to add a mod to the download list without going inside.
Feature: Check if the mod is already installed.

Feature: Instead of forcing the code (i.e. 33294435) to be the name of the folder so updates can be downloaded from steam, make it so it only check the code up to the first space so mods folders can be named whatever for human readibility. Noticed that folder scheme/name is all the app uses to convert mods to Steam from local. Make the symlink follow your convention but the download be named as mod title.

Feature: Clean up the [br] (bbcode) coding on descriptions.

Clarity: Textures - delete .dds to Textures - clear optimizations.
Clarity: change the name of "download missing mods automatically" to what it actually does (i.e. search and setup the download for missing mods without forcing them to be downloaded) - this behavior is great to me so if you decide to ever make proper auto download keep the "Check for missing mods and prepare download" too.

Problem: Your folder inspector daemon reinstall mods or the folder code at least after some time when the mod has been deleted by the user or by the app.
Solution: This is a bug, but yes. In my other report the daemon causes crashes when too many changes occur, when you try to delete a mod that is actively being used by the game and also does this. I like your daemon, but had to disable it due to this 3 things.

Feature; Meanwhile more in app sorting options like "Last download/last installed" pop up, it would be good if last added mods could be at the bottom or something in the inactive list besides the alphabetical order.

Great app by the way, this long post is not a critique of what it is missing according to me, but the opposite that I have used it SO MUCH that I noticed what could make it better. Sounds weird, is weird, but I'm thankful.

**Comment by Raizanad** [comment]
Since there is no easy way of moving mods from folder A to folder B (and I like the direct deletion method instead of trash bin though that could have been a band aid fix for my following feature request):
Feature: Move selected mods to folder.
Probably between a previously defined folder A and B in settings, as to migrate inactive mods across where the game does not read them but keep an eye for mods that are waiting to be updated for example.

**Comment by oceancabbage** [comment]
Thank you very much for using RimSort and also for the detailed suggestions/improvements :) if it's okay with you, let's keep this issue open and developers (including myself) can discuss and work through the most impactful features here. In the meantime, feel free to keep adding to this issue anything you would like to see in a future RimSort version.

**Comment by Raizanad** [comment]
Yes, no problem. I am considering making individual feature request/bug report with each one for better tracking (as it is meant to be done) as I keep getting more and more invested into RimSort because I have been watching the energy getting poured into it. Assuming that is desirable that is.

**Comment by Raizanad** [comment]
- Check if mod is still available in Steam before deleting it when updating.
- Add the option to keep a backup of the previous version of a mod when updating (while allowing to clear cache, or to pick which mods will have this option) - this can be done easily by just moving a "backed" old version to a "backup" or "versions" folder.
- A changelog of the last added/removed mods from your mod list, but more importantly, a changelog of the mods that have been updated. Generally updates should be positive, but I can see disastrous events due to mod updating.

And not important but maybe add a XML parser that adds:
- Checking files for the <?xml> start existing.
- And for general xml malformations.

- Clear languages/versions from downloaded mods as defined by user: i.e. remove every other language translation that isn't English, and every folder that isn't 1.5.

**Comment by Chunnyluny** [comment]
> * 1 Check if mod is still available in Steam before deleting it when updating.

> * 2 Add the option to keep a backup of the previous version of a mod when updating (while allowing to clear cache, or to pick which mods will have this option) - this can be done easily by just moving a "backed" old version to a "backup" or "versions" folder.

- [ ] Add a check to see if a mod is still accessible on Steam, if not make a back-up copy of the mod folder that shouldn't be deleted or modified to not break save games. 
This task combines 1+2 together. 

- [ ] Make it possible to put mods in quarantine/in a vault until they are updated, when you remove the outdated mod, you forget to find an alternative.
Otherwise, make it possible to have V1.4 and V1.5 mods (this is the same as the instances but it is still a bit weird to use them)

> * 3 A changelog of the last added/removed mods from your mod list, but more importantly, a changelog of the mods that have been updated. Generally updates should be positive, but I can see disastrous events due to mod updating.

- [x] Make an Update log feature like in RimPy but to make it more clear, use filter names where you can click on those names, like you can do in windows file explorer, for example; on date modified, on date changed, on name, on file size, on install location. 

- Make an Update log feature like in RimPy, in RimPy they use green, red and white text to show the state of a mod, green means active in the mod list, red means inactive, or removed from your pc, otherwise moved to another folder on your PC or on another drive. 

- Another suggestion is to get the mod that Mlie made, Show mod updates, it makes it easier for players to see what mods have been updated since the last time they played their savegame. This mod makes it easier for you to find out which mods have been updated since the last time you played, for example when trying to find out what mod has started causing an issue. Mod is found here: https://steamcommunity.com/sharedfiles/filedetails/?id=3218690663

> * 4 A XML parser that checks for XML prolog existing, and warns users about malformations. 

- [ ] Add a XML parser that checks for XML malformations and warns the user that it needs to report this to the mod author. If there are malformations in the About.xml it can cause the mod to not work, for example; no closing tags, or <description> <description> (/ is missing in end tag)

- A file that misses <?xml version="1.0" encoding="utf-8"?> can't be parsed, if possible, there should be a warning which says this mod contains malformations or has no XML prolog.
To avoid errors, you should specify the encoding used, or save your XML files as UTF-8.

> * Clear languages/versions from downloaded mods as defined by user: i.e. remove every other language translation that isn't English, and every folder that isn't 1.5.

- ?? This suggestion doesn't make sense to me, the language folders don't take up a lot of space for the mod, and it can only cause issues when folders are removed and modified. 

> Posted this in discord to highlight them as priority, double dumping here as to not risk the request getting buried forever in discord. 
- Note1: good that you have posted them as a comment as well, so we can keep all things organised.

- Note2: I edited the post to make it more like a todo list with checkboxes for easier task management like I said earlier in discord.

**Comment by Chunnyluny** [comment]
> Since there is no easy way of moving mods from folder A to folder B (and I like the direct deletion method instead of trash bin though that could have been a band aid fix for my following feature request): Feature: Move selected mods to folder. Probably between a previously defined folder A and B in settings, as to migrate inactive mods across where the game does not read them but keep an eye for mods that are waiting to be updated for example.

This looks like my suggestion of a vault/putting mods in quarantine (on your later feature request)

**Comment by Raizanad** [comment]
>         Clear languages/versions from downloaded mods as defined by user: i.e. remove every other language translation that isn't English, and every folder that isn't 1.5.
> 
>     ?? This suggestion doesn't make sense to me, the language folders don't take up a lot of space for the mod, and it can only cause issues when folders are removed and modified.

Yes, it may not make sense at all. The suggestion popped in the first place while I was checking for malformed code in the mods I have downloaded and noticed some translation in languages I don't use (i.e. Korean for x mod had malformed code, like for example not having the starting <LanguageData> in some instances). Since translations may not be monitored by a modder it may cause issues (or not at all, since I'm not sure if the game even reads/cares - even if there is malformed code - in unused language files) but I can't tell since the game startup is all but logical in some ways.

**Comment by hk21702** [comment]
> 4 A XML parser that checks for XML prolog existing, and warns users about malformations.
This is already in the works with my metadata rework. Though it will take time and there are currently blockers before I can continue with it. Don't expect this to be completed for a good bit since this is a large task.

> Clear languages/versions from downloaded mods as defined by user: i.e. remove every other language translation that isn't English, and every folder that isn't 1.5.
The language stuff was already touched on in a previous comment. There are dangers to removing all folders apart from a specific version folder because some mod authors organize things in a way such that some folders are version-agnostic and are loaded regardless of what version. Some also have special conditionals for loading folders that may be dependent on other mods in the load order. It's simply too dangerous to be removing data from mods.
